### ISSUE TITLE
Weird error involving potential `Missing` in union split

### ISSUE THREAD

**Issue by gdalle** [aviatesk/JET.jl#541]
Hi there,
I'm currently trying to get Graphs.jl JET-approved, and I only have **one** error to fix.
```julia
   ═════ 2 possible errors found ═════
  ┌ ==(g::Graphs.SimpleGraphs.SimpleDiGraph, h::Graphs.SimpleGraphs.SimpleDiGraph) @ Graphs.SimpleGraphs /home/runner/work/Graphs.jl/Graphs.jl/src/SimpleGraphs/simpledigraph.jl:411
  │ non-boolean `Missing` found in boolean context (1/2 union split): goto %18 if not (Graphs.SimpleGraphs.fadj(g::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer Graphs.SimpleGraphs.:(==) Graphs.SimpleGraphs.fadj(h::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer)::Union{Missing, Bool}
  └────────────────────
  ┌ ==(g::Graphs.SimpleGraphs.SimpleDiGraph, h::Graphs.SimpleGraphs.SimpleDiGraph) @ Graphs.SimpleGraphs /home/runner/work/Graphs.jl/Graphs.jl/src/SimpleGraphs/simpledigraph.jl:411
  │ non-boolean `Missing` found in boolean context (1/2 union split): goto %18 if not (Graphs.SimpleGraphs.badj(g::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer Graphs.SimpleGraphs.:(==) Graphs.SimpleGraphs.badj(h::Graphs.SimpleGraphs.SimpleDiGraph)::Array{Vector{T}, 1} where T<:Integer)::Union{Missing, Bool}
  └────────────────────
```
Basically I'm comparing two adjacency lists with `fadj(g) == fadj(h)` and JET is afrait that the result might be `Missing`, but I really don't know why.
See the CI log: https://github.com/JuliaGraphs/Graphs.jl/actions/runs/5344089300/jobs/9688032589?pr=249

**Comment by gdalle** [comment]
More precisely, this is the guilty function:
```julia
function ==(g::SimpleDiGraph{T1}, h::SimpleDiGraph{T2}) where {T1,T2}
    if vertices(g) == vertices(h) && ne(g) == ne(h)
        if fadj(g) == fadj(h) && badj(g) == badj(h)  # fails JET
        # if all(fadj(g) .== fadj(h)) && all(badj(g) .== badj(h))  # passes JET
            return true
        else
            return false
        end
    else
        return false
    end
end
```
The results of `fadj(g)` and `badj(g)` are adjacency lists, ie. `Vector{Vector{<:Integer}}`. So it seems comparing these objects directly fails, but I just found that if I compare the nested vectors individually it passes.

**Comment by timholy** [comment]
> I'm currently trying to get Graphs.jl JET-approved

Tangent here, but what exactly does that mean? `report_package`? If Graphs has a precompile workload it would be interesting to compare the `MethodAnalysis`-based approach in https://aviatesk.github.io/JET.jl/dev/tutorial/#Analyze-packages-using-a-representative-workload and see if it detects any other issues. `report_package` uses abstract inference, and the analysis stops at any runtime dispatch (https://aviatesk.github.io/JET.jl/stable/#Limitations), so there's a chance you'd discover different things if you're using concrete MethodInstance signatures.

**Comment by gdalle** [comment]
> Tangent here, but what exactly does that mean? report_package?

Yes, or in this case `test_package`

> If Graphs has a precompile workload it would be interesting to compare the MethodAnalysis-based approach

It doesn't yet have one, but I think that might be interesting to put in place as soon as GraphsBase.jl emerges.

> report_package uses abstract inference, and the analysis stops at any runtime dispatch

Indeed, and I seriously doubt all the code in Graphs.jl is inferrable. But JET already caught quite a few mistakes, so abstract interpretation is already better than nothing!

**Comment by timholy** [comment]
> abstract interpretation is already better than nothing

Indeed! I don't think there's a single solution to all these problems, but the main realistic possibilities are being filled out.

**Comment by jakobnissen** [comment]
This is not an issue with JET. Here is a minimal example:

```julia
julia> f() = Vector{Any}([1,2,3])
f (generic function with 1 method)

julia> function foo()
           if f() == f()
               1    
           else
               2
           end
       end

foo (generic function with 1 method)

julia> using JET

julia> @report_call foo()
═════ 1 possible error found ═════
┌ foo() @ Main ./REPL[2]:2
│ non-boolean `Missing` found in boolean context (1/2 union split): goto %6 if not (f()::Vector{Any} == f()::Vector{Any})::Union{Missing, Bool}
```

If brief, the issue is that `==(::Vector, ::Vector)`, when the element type of the vectors are unknown, is inferred to be `Union{Bool, Missing}`. This inference is reasonable given the definition of `==` on vectors. You then take this output and put it into an if-statement, which will fail if the `==` returns `missing`.

The solution is to make `fadj` and `badj` well-inferred.

**Comment by jakobnissen** [comment]
Perhaps you are testing digraphs with abstract types? In that case, it's hard to avoid potential issues.

```julia
julia> report_call(==, (SimpleDiGraph{Int}, SimpleDiGraph{Int}))
No errors detected

julia> report_call(==, (SimpleDiGraph{Integer}, SimpleDiGraph{Integer}))
═════ 1 possible error found ═════
┌ ==(g::SimpleDiGraph{Integer}, h::SimpleDiGraph{Integer}) @ Graphs.SimpleGraphs /home/jakni/.julia/packages/Graphs/7SMZs/src/SimpleGraphs/simpledigraph.jl:383
│ non-boolean `Missing` found in boolean context (1/2 union split): goto %17 if not (Graphs.SimpleGraphs.fadj(g::SimpleDiGraph{Integer})::Vector{Vector{Integer}} Graphs.SimpleGraphs.:(==) Graphs.SimpleGraphs.fadj(h::SimpleDiGraph{Integer})::Vector{Vector{Integer}})::Union{Missing, Bool}
```

**Comment by aviatesk** [comment]
Following PR #543, JET ignores the possibility of a poorly-inferred `x == y` call returning `missing` during the `report_package` analysis. Refer to issue #542 for reasons justifying this behavior. Essentially, `report_package` often relies on poor input argument type information at the beginning of analysis, leading to noisy error reports for function definitions like:

```julia
struct MyToken end
ismytoken(x) = x == MyToken()
```

This error is arguably just noise when the target package does not handle `missing`. `report_package` is designed as an entry point for easy analysis, even at the cost of accuracy, so it is not sound from the beginning. Hence, it might be beneficial to simply ignore such noise. A package that handles `missing` can disable this behavior if necessary.

However, in interactive entry points like `report_call`, where concrete input argument types are available, this behavior should be turned off. This is because, if the code, when given specific input argument types, results in a `Union{Bool,Missing}` possibility, it likely signifies an inferrability issue or the code really needs to handle `missing`.

**Comment by gdalle** [comment]
Oops, the original bug is fixed by JET 0.8.2 on Julia 1.9, but a new similar one appeared with Julia 1.10

https://github.com/JuliaGraphs/Graphs.jl/actions/runs/5377914717/jobs/9757016398

**Comment by gdalle** [comment]
Although in this case I think the fault lies with bad inference in Graphs.jl

**Comment by aviatesk** [comment]
> Oops, the original bug is fixed by JET 0.8.2 on Julia 1.9, but a new similar one appeared with Julia 1.10
> 
> https://github.com/JuliaGraphs/Graphs.jl/actions/runs/5377914717/jobs/9757016398

It should be fixed by JET 0.8.3, which is waiting for https://github.com/JuliaRegistries/General/pull/86248 to be merged.
